README.md
# NAME

Plack::Middleware::REST - Route PSGI requests for RESTful web applications

# STATUS

[![Build Status](https://travis-ci.org/nichtich/Plack-Middleware-REST.png)](https://travis-ci.org/nichtich/Plack-Middleware-REST)
[![Coverage Status](https://coveralls.io/repos/nichtich/Plack-Middleware-REST/badge.png)](https://coveralls.io/r/nichtich/Plack-Middleware-REST)
[![Kwalitee Score](http://cpants.cpanauthors.org/dist/Plack-Middleware-REST.png)](http://cpants.cpanauthors.org/dist/Plack-Middleware-REST)

# SYNOPSIS

    # $get, $update, $delete, $create, $list, $patch, $app must be PSGI applications
    builder {
        enable 'REST',
            get          => $get,           # GET /{id}
            upsert       => $update,        # PUT /{id}
            delete       => $delete,        # DELETE /{id}
            create       => $create,        # POST /
            list         => $list,          # GET /
            patch        => $patch,         # PATCH /{id}
            head         => 1,              # HEAD /{$id} => $get, HEAD / => $list
            options      => 1,              # support OPTIONS requests
            pass_through => 1,              # pass everything else to $app
            patch_types  => ['text/plain']; # optional accepted patch types
        $app;
    };

# DESCRIPTION

Plack::Middleware::REST routes HTTP requests (given in [PSGI](https://metacpan.org/pod/PSGI) request format)
on the principles of Representational State Transfer (REST). In short, the
application manages a set of resources with common base URL, each identified by
its URL. One can retrieve, create, update, delete, list, and patch resources
based on HTTP request methods.

Let's say an instance of Plack::Middleware::REST is mounted at the base URL
`http://example.org/item/`. The following HTTP request types can be
recognized, once they [have been assigned](#configuration):

- `POST http://example.org/item/`

    Calls the PSGI application `create` to create a new resource with URL assigned
    by the application.

- `GET http://example.org/item/123`

    Calls the application `get` to retrieve an existing resource identified by
    `http://example.org/item/123`.

- `PUT http://example.org/item/123`

    Calls the PSGI application `upsert` to either update an existing resource
    identified by `http://example.org/item/123` or to create a new resource with
    this URL. The application may reject updates and/or creation of new resources,
    acting like an update or insert method.

- `DELETE http://example.org/item/123`

    Calls the PSGI application `delete` to delete an existing resource identified
    by `http://example.org/item/123`.

- `GET http://example.org/item/`

    Calls the PSGI application `list` to get a list of existing resources.

- `PATCH http://example.org/item/123`

    Calls the PSGI application `patch` to update an existing resource
    identified by `http://example.org/item/123`. The application may
    reject updates of resources.

- `OPTIONS http://example.org/item/`

    Calls the PSGI application to return the allowed methods for the resource.

Other requests result either result in a PSGI response with error code 405 and
a list of possible request types in the `Accept` header, or the request is
passed to the underlying application in the middleware stack, if option
`pass_through` is set.

# CONFIGURATION

## get

## create

## upsert

## delete

## list

## patch

The options `get`, `create`, `upsert`, `delete`, `list`, `patch` can be set
to PSGI applications to enable the corresponding REST request type. One can also
use string aliases, including `app` to pass the request in the middleware stack:

    builder {
        enable 'REST',
            get          => 'app',   # pass GET requests on resource to $wrapped
            create       => $create, # pass POST to base URL to $create
            upsert       => $update; # pass PUT requests on resources to $update
            pass_through => 0;       # respond other requests with 405
        $wrapped;
    };

## head

By default (`head => 1`) the app configured to `get` and/or `list` resources
are also assumed to handle HEAD requests. Setting this configuration to `0` will
disallow HEAD requests. The special value `auto` will rewrite HEAD requests with
[Plack::Middleware::Head](https://metacpan.org/pod/Plack::Middleware::Head).

## options

By default (`options => 1`) the app is configured to handle OPTIONS requests
for a resource. Setting this configuration to `0` will dissallow OPTIONS requests.

## pass\_through

Respond to not allowed requests with HTTP 405. Enabled by default, but this may
change in a future version of this module!

## patch\_types

Optional array of acceptable patch document types for PATCH requests.
Respond to unacceptable patch document types with HTTP 415.

# COPYRIGHT AND LICENSE

Copyright 2014- Jakob Voß

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

# CONTRIBUTORS

Jakob Voß and Chris Kirke

# SEE ALSO

- [Plack::Middleware::REST::Util](https://metacpan.org/pod/Plack::Middleware::REST::Util), included with Plack::Middleware::REST
provides some utility methods to implement RESTful PSGI applications.  The
module may be removed in a future release.
- See [Plack::Middleware::Negotiate](https://metacpan.org/pod/Plack::Middleware::Negotiate) for content negotiation.
- See [Plack::Middleware::ETag](https://metacpan.org/pod/Plack::Middleware::ETag) for ETag generation.
- Alternative CPAN modules with similar scope include [Apache2::REST](https://metacpan.org/pod/Apache2::REST),
[REST::Utils](https://metacpan.org/pod/REST::Utils), [REST::Application](https://metacpan.org/pod/REST::Application), [WWW::REST::Apid](https://metacpan.org/pod/WWW::REST::Apid), [WWW::REST::Simple](https://metacpan.org/pod/WWW::REST::Simple),
[CGI::Application::Plugin::REST](https://metacpan.org/pod/CGI::Application::Plugin::REST), and [Plack::App::REST](https://metacpan.org/pod/Plack::App::REST).  Moreover
there are general web application frameworks like [Dancer](https://metacpan.org/pod/Dancer)/[Dancer2](https://metacpan.org/pod/Dancer2),
[Mojolicious](https://metacpan.org/pod/Mojolicious), and [Catalyst](https://metacpan.org/pod/Catalyst). Maybe the number of such modules and
frameworks is higher than the number of actual web APIs written in Perl. Who
knows?
- REST client modules at CPAN include [REST::Client](https://metacpan.org/pod/REST::Client), [Eixo::Rest](https://metacpan.org/pod/Eixo::Rest),
[REST::Consumer](https://metacpan.org/pod/REST::Consumer), [Net::Rest::Generic](https://metacpan.org/pod/Net::Rest::Generic), [LWP::Simple::REST](https://metacpan.org/pod/LWP::Simple::REST), and
[WWW:.REST](WWW:.REST), [Role::REST::Client](https://metacpan.org/pod/Role::REST::Client), [Rest::Client::Builder](https://metacpan.org/pod/Rest::Client::Builder),
[MooseX::Role::REST::Consumer](https://metacpan.org/pod/MooseX::Role::REST::Consumer). Don't ask why.


Changes
Changelog for Plakc::Middleware::REST. See also http://github.com/nichtich/Plack-Middleware-REST/issues

{{$NEXT}}

0.10  2015-04-06 09:11:57 CEST
  - support HEAD, OPTIONS, and PATCH requests (Chris Kirke)


t/lib/RESTApp.pm
package RESTApp;
use strict;
use warnings;

use Plack::Request;
use Plack::Middleware::REST::Util;
use HTTP::Status qw(status_message);

sub new { 
    bless { hash => {}, count => 0 }, shift; 
}

## helper method

sub response {
    my $code = shift;
    my $body = @_ ? shift : status_message($code); 
    [ $code, [ 'Content-Type' => 'text/plain', @_ ], [ $body ] ];
}

## methods

sub resource {
    my ($self,$env) = @_;
    $self->{hash}->{ request_id($env) };
}

sub get {
    my ($self,$env) = @_;
    my $resource = $self->resource($env);
    return response(404) unless defined $resource;
    return response( 200 => ($env->{REQUEST_METHOD} eq 'GET' ? $resource : ''));
}

sub create {
    my ($self,$env) = @_;
    my ($resource, $type) = request_content($env);
    return response(400) unless defined $resource;

    my $id = ++$self->{count};
    $self->{hash}->{ $id } = $resource;

    my $uri = request_uri($env,$id);
    return response(201, $resource, Location => $uri); # or 204
}

sub update {
    my ($self,$env) = @_;
    
    return response(404) unless defined $self->resource($env);

    my ($resource, $type) = request_content($env);
    return response(400) unless defined $resource;

    $self->{hash}->{ request_id($env) } = $resource;
    return response(200,$resource); # or 204
}

sub delete {
    my ($self,$env) = @_;
    return (defined (delete $self->{hash}->{ request_id($env) })) 
        ? response(204,'') : response(404);
}

sub list {
    my ($self,$env) = @_;
    my @uris = map { request_uri($env,$_) } sort keys %{$self->{hash}}; 
    return response( 200 => '' ) if $env->{REQUEST_METHOD} eq 'HEAD';
    response(200, join "\n", @uris);
}

1;


lib/Plack/Middleware/REST/Util.pm
package Plack::Middleware::REST::Util;
#ABSTRACT: Utility methods to create RESTful PSGI applications

our $VERSION = '0.03';

use Plack::Request;

use parent 'Exporter';

our @EXPORT = qw(request_id request_content request_uri);

sub request_id {
    substr($_[0]->{PATH_INFO} || '/',1)
}

sub request_content {
    return unless defined $_[0]->{CONTENT_LENGTH};
    return (Plack::Request->new($_[0])->content => $_[0]->{CONTENT_TYPE});
}

sub request_uri {
    my $env = shift;
    my $id  = @_ ? shift : request_id($env);
    my $uri = Plack::Request->new($env)->base;
    $uri .= '/' unless $uri =~ qr{/$};
    return $uri . $id;
}

1;

=head1 SYNOPSIS

    use Plack::Middleware::REST::Util;

    $id  = request_id($env)      # empty string or local resource identifier
    $uri = request_uri($env)     # resource identifier of current request
    $uri = request_uri($env,$id) # resource identifier of modified request
    ($content => $type) = request_content($env); # send content and MIME type

=cut


lib/Plack/Middleware/REST.pm
package Plack::Middleware::REST;
use strict;
use warnings;

our $VERSION = '0.03';

use Carp qw(croak);
use Scalar::Util qw(reftype);

use parent 'Plack::Middleware';
use Plack::Util::Accessor qw(get create upsert delete list head pass_through routes options patch patch_types );

use Plack::Middleware::Head;

sub prepare_app {
    my ($self) = @_;

    $self->pass_through(0) unless defined $self->pass_through;
    $self->head(1) unless defined $self->head;
    $self->options(1) unless defined $self->options;

    $self->routes({
        resource   => {
            GET    => 'get',
            PUT    => 'upsert',
            DELETE => 'delete',
            PATCH  => 'patch',
        },
        collection => {
            GET    => 'list',
            POST   => 'create',
        },
    });

    if ($self->head) {
        $self->routes->{resource}->{HEAD} = 'get';
        $self->routes->{collection}->{HEAD} = 'list';
    }

    foreach my $action (qw(get create upsert delete list patch))  {
        my $app = $self->{$action};

        # alias
        $self->{$action} = $self->{$app} if $app and !ref $app;

        croak "PSGI application '$action' must be code reference"
            if $self->{action} and (reftype($self->{$action}) || '') ne 'CODE';
    }

    while (my ($type, $route) = each %{$self->routes}) {
        $self->{allow}->{$type} = join ', ',
            sort grep { $self->{ $route->{$_} } } keys %$route;
        foreach my $method (keys %$route) {
            $route->{$method} = $self->{ $route->{$method} };
        }
        if ($self->head eq 'auto') {
            $route->{HEAD} = Plack::Middleware::Head->wrap($route->{HEAD});
        }
    }
}

sub call {
    my ($self, $env) = @_;

    my $type   = ($env->{PATH_INFO} || '/') eq '/' ? 'collection' : 'resource';
    my $method = $env->{REQUEST_METHOD};

    if ($method eq 'OPTIONS') {
        if ($self->options) {
            my %headers = ( 'Allow' => $self->{allow}->{$type} );
            if ($self->patch() && $self->patch_types()) {
                $headers{'Accept-Patch'} = join( ',', @{$self->patch_types()} );
            }
            [ 200, [ %headers ], [] ];
        } else {
            [ 405, [ Allow => $self->{allow}->{$type} ], ['Method Not Allowed'] ];
        }
    } else {
        my $app    = $self->routes->{$type}->{$method};
        $app ||= $self->{app} if $self->pass_through;
        if ( $app ) {
            if (($method eq 'PATCH') && ($self->patch_types())
                    && ! grep { /^$env->{'CONTENT_TYPE'}$/msx } @{$self->patch_types()} ) {
                [ 415, [], ['Unknown Patch Type'] ];
            } else {
                $app->($env);
            }
        } else {
            [ 405, [ Allow => $self->{allow}->{$type} ], ['Method Not Allowed'] ];
        }
    }
}

1;
__END__

=encoding utf8

=head1 NAME

Plack::Middleware::REST - Route PSGI requests for RESTful web applications

=begin markdown

# STATUS

[![Build Status](https://travis-ci.org/nichtich/Plack-Middleware-REST.png)](https://travis-ci.org/nichtich/Plack-Middleware-REST)
[![Coverage Status](https://coveralls.io/repos/nichtich/Plack-Middleware-REST/badge.png)](https://coveralls.io/r/nichtich/Plack-Middleware-REST)
[![Kwalitee Score](http://cpants.cpanauthors.org/dist/Plack-Middleware-REST.png)](http://cpants.cpanauthors.org/dist/Plack-Middleware-REST)

=end markdown

=head1 SYNOPSIS

    # $get, $update, $delete, $create, $list, $patch, $app must be PSGI applications
    builder {
        enable 'REST',
            get          => $get,           # GET /{id}
            upsert       => $update,        # PUT /{id}
            delete       => $delete,        # DELETE /{id}
            create       => $create,        # POST /
            list         => $list,          # GET /
            patch        => $patch,         # PATCH /{id}
            head         => 1,              # HEAD /{$id} => $get, HEAD / => $list
            options      => 1,              # support OPTIONS requests
            pass_through => 1,              # pass everything else to $app
            patch_types  => ['text/plain']; # optional accepted patch types
        $app;
    };

=head1 DESCRIPTION

Plack::Middleware::REST routes HTTP requests (given in L<PSGI> request format)
on the principles of Representational State Transfer (REST). In short, the
application manages a set of resources with common base URL, each identified by
its URL. One can retrieve, create, update, delete, list, and patch resources
based on HTTP request methods.

Let's say an instance of Plack::Middleware::REST is mounted at the base URL
C<http://example.org/item/>. The following HTTP request types can be
recognized, once they L<have been assigned|/CONFIGURATION>:

=over 4

=item C<POST http://example.org/item/>

Calls the PSGI application C<create> to create a new resource with URL assigned
by the application.

=item C<GET http://example.org/item/123>

Calls the application C<get> to retrieve an existing resource identified by
C<http://example.org/item/123>.

=item C<PUT http://example.org/item/123>

Calls the PSGI application C<upsert> to either update an existing resource
identified by C<http://example.org/item/123> or to create a new resource with
this URL. The application may reject updates and/or creation of new resources,
acting like an update or insert method.

=item C<DELETE http://example.org/item/123>

Calls the PSGI application C<delete> to delete an existing resource identified
by C<http://example.org/item/123>.

=item C<GET http://example.org/item/>

Calls the PSGI application C<list> to get a list of existing resources.

=item C<PATCH http://example.org/item/123>

Calls the PSGI application C<patch> to update an existing resource
identified by C<http://example.org/item/123>. The application may
reject updates of resources.

=item C<OPTIONS http://example.org/item/>

Calls the PSGI application to return the allowed methods for the resource.

=back

Other requests result either result in a PSGI response with error code 405 and
a list of possible request types in the C<Accept> header, or the request is
passed to the underlying application in the middleware stack, if option
C<pass_through> is set.

=head1 CONFIGURATION

=head2 get

=head2 create

=head2 upsert

=head2 delete

=head2 list

=head2 patch

The options C<get>, C<create>, C<upsert>, C<delete>, C<list>, C<patch> can be set
to PSGI applications to enable the corresponding REST request type. One can also
use string aliases, including C<app> to pass the request in the middleware stack:

    builder {
        enable 'REST',
            get          => 'app',   # pass GET requests on resource to $wrapped
            create       => $create, # pass POST to base URL to $create
            upsert       => $update; # pass PUT requests on resources to $update
            pass_through => 0;       # respond other requests with 405
        $wrapped;
    };

=head2 head

By default (C<head =E<gt> 1>) the app configured to C<get> and/or C<list> resources
are also assumed to handle HEAD requests. Setting this configuration to C<0> will
disallow HEAD requests. The special value C<auto> will rewrite HEAD requests with
L<Plack::Middleware::Head>.

=head2 options

By default (C<options =E<gt> 1>) the app is configured to handle OPTIONS requests
for a resource. Setting this configuration to C<0> will dissallow OPTIONS requests.

=head2 pass_through

Respond to not allowed requests with HTTP 405. Enabled by default, but this may
change in a future version of this module!

=head2 patch_types

Optional array of acceptable patch document types for PATCH requests.
Respond to unacceptable patch document types with HTTP 415.

=head1 COPYRIGHT AND LICENSE

Copyright 2014- Jakob Voß

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=head1 CONTRIBUTORS

Jakob Voß and Chris Kirke

=head1 SEE ALSO

=over

=item

L<Plack::Middleware::REST::Util>, included with Plack::Middleware::REST
provides some utility methods to implement RESTful PSGI applications.

=item

See L<Plack::Middleware::Negotiate> for content negotiation.

=item

See L<Plack::Middleware::ETag> for ETag generation.

=item

Alternative CPAN modules with similar scope include L<Apache2::REST>,
L<REST::Utils>, L<REST::Application>, L<WWW::REST::Apid>, L<WWW::REST::Simple>,
L<CGI::Application::Plugin::REST>, and L<Plack::Middleware::RestAPI>.  Moreover
there are general web application frameworks like L<Dancer>/L<Dancer2>,
L<Mojolicious>, and L<Catalyst>. Maybe the number of such modules and
frameworks is higher than the number of actual web APIs written in Perl. Who
knows?

=item

REST client modules at CPAN include L<REST::Client>, L<Eixo::Rest>,
L<REST::Consumer>, L<Net::Rest::Generic>, L<LWP::Simple::REST>, and
L<WWW:.REST>, L<Role::REST::Client>, L<Rest::Client::Builder>,
L<MooseX::Role::REST::Consumer>. Don't ask why.

=back

=cut
